Sources/GRPC/GRPCClientStateMachine.swift

```diff
@@ -0,0 +1,588 @@
+  /// - `.clientIdleServerIdle`
+  /// - `.clientStreamingServerStreaming`
+  /// - `.clientClosedServerStreaming`
+  /// - `.clientClosedServerClosed`
+  ///
+  /// - Parameter headers: The headers recieved from the server.
+  mutating func recieveResponseHeaders(
+    _ responseHead: HTTPResponseHead
+  ) -> Result<HTTPHeaders, RecieveResponseHeadError> {
+    return self.state.recieveResponseHeaders(responseHead, logger: self.logger)
+  }
+
+  /// Read a response buffer from the server and return any decoded messages.
+  ///
+  /// If the response stream has an expected count of `.one` then this function is guaranteed to
+  /// produce *at most* one `Response` in the `Result`.
+  ///
+  /// To recieve a response buffer the server must be streaming. Valid states are:
+  /// - `.clientClosedServerStreaming` → `.clientClosedServerStreaming`
+  /// - `.clientStreamingServerStreaming` → `.clientStreamingServerStreaming`
+  ///
+  /// It is invalid and fatal to recieve a response in the following states:
+  /// - `.clientIdleServerIdle`
+  /// - `.clientClosedServerStreaming`
+  /// - `.clientStreamingServerStreaming`
+  /// - `.clientClosedServerClosed`
+  ///
+  /// - Parameter buffer: A buffer of bytes received from the server.
+  mutating func recieveResponse(
+    _ buffer: inout ByteBuffer
+  ) -> Result<[Response], MessageReadError> {
+    return self.state.recieveResponse(&buffer)
+  }
+
+  /// Recieve the end of the response stream from the server and parse the results into
+  /// a `GRPCStatus`.
+  ///
+  /// To close the response stream the server must be streaming or idle (since the server may choose
+  /// to 'fast fail' the RPC). Valid states are:
+  /// - `.clientStreamingServerIdle` → `.clientClosedServerClosed`
+  /// - `.clientStreamingServerStreaming` → `.clientClosedServerClosed`
+  /// - `.clientClosedServerIdle` → `.clientClosedServerClosed`
+  /// - `.clientClosedServerStreaming` → `.clientClosedServerClosed`
+  ///
+  /// It is invalid to recieve an end-of-stream if the RPC has not been initiated or has already
+  /// been terminated. That is, in one of the following states:
+  /// - `.clientIdleServerIdle`
+  /// - `.clientClosedServerClosed`
+  ///
+  /// - Parameter trailers: The trailers to parse.
+  mutating func recieveEndOfResponseStream(
+    _ trailers: HTTPHeaders
+  ) -> Result<GRPCStatus, InvalidStateError> {
+    return self.state.recieveEndOfResponseStream(trailers)
+  }
+}
+
+extension GRPCClientStateMachine.State {
+  /// See `GRPCClientStateMachine.sendRequestHeaders(host:path:options:requestID)`.
+  mutating func sendRequestHeaders(
+    host: String,
+    path: String,
+    options: CallOptions,
+    requestID: String
+  ) -> Result<HTTPRequestHead, InvalidStateError> {
+    let result: Result<HTTPRequestHead, InvalidStateError>
+
+    switch self {
+    case let .clientIdleServerIdle(pendingWriteState, responseArity):
```

Comment:
Should we always call this `{Request,Response}Arity` instead of `{Request,Response}Count`, both the variable names and the type name?

Reply:
We should be consistent, yes. I originally had it as `MessageArity` but renamed to `MessageCount` since it's a "more obvious" (for want of a better term) name.

Reply:
Has this been done? (Sorry, don't have the time for a full re-review right now.)

Reply:
Yes: it's now `MessageArity` with an additional case for `none`.